### DISCUSSION TITLE
Window Snap Functionality Fails in Stage Manager Mode Post MacOS Update

### DISCUSSION THREAD

**Discussion by PANXIONG-CN** [rxhanson/Rectangle#1555] [General]
I am currently encountering an issue with the Rectangle app after upgrading to the latest MacOS. In Stage Manager mode, when I try to drag a window to the left, it doesn't snap and split screen as it's supposed to. Instead, it gets incorporated into the Stage Manager interface.

This issue seems to be consistently happening whenever I attempt to drag and snap windows. It's quite disruptive to my workflow, especially as I heavily rely on the split-screen functionality for multitasking.

If there's any additional information or debugging data you need to help resolve this issue, please let me know and I will do my best to provide it.

Thank you for your attention to this matter.

**Comment by rxhanson** [comment]
Thanks for reporting. I am not able to reproduce this, except for with windows that I pull out of stage manager then move back in without dropping. Perhaps a screen recording would help? You can also view logs by opening the menu bar menu, holding option, and selecting "View Logging..."

**Reply by rxhanson** [reply]
I've been able to reproduce this with some of the reports in this discussion. Unfortunately, as far as I can tell, this is an issue with how macOS handles dragging and dropping in the Stage Manager, and it will override anything that a third party such as Rectangle can do. With enough people filing an issue to Apple, there's always potential that they will change the way that this works to accommodate 3rd party window management with Stage Manager.

**Comment by ubuntudroid** [comment]
I found this mostly happens when having multiple windows open. Logs stay empty.

**Comment by ubuntudroid** [comment]
Ah, I should add that I am on an external monitor with the connected MacBook's lid being closed. Maybe that helps with reproduction?

**Comment by jpatrolla** [comment]
edit:
also, make sure if you've just installed pro that your original rectangle instance isn't running lol

--

Hi gang. Had a similar problem. Try changing the following setting to 'Do Nothing':

![image](https://github.com/rxhanson/Rectangle/assets/9730683/987b3465-a620-4f47-b2d7-3820c5730148)

System Settings > Desktop & Dock > Double-click a window's title bar to **Do Nothing**

Seems to be fixed so far for me. Hope that helps

**Comment by weekijoon** [comment]
> edit: also, make sure if you've just installed pro that your original rectangle instance isn't running lol
> 
> --
> 
> Hi gang. Had a similar problem. Try changing the following setting to 'Do Nothing':
> 
> ![image](https://user-images.githubusercontent.com/9730683/277159039-987b3465-a620-4f47-b2d7-3820c5730148.png)
> 
> System Settings > Desktop & Dock > Double-click a window's title bar to **Do Nothing**
> 
> Seems to be fixed so far for me. Hope that helps

This fixed snapping not working for me after trying a few other suggestions. Thanks!

**Comment by mischa-hildebrand** [comment]
I can reproduce this issue reliably. For me, it only occurs when having multiple windows grouped and then trying to move one of those windows to the left edge of the screen. It always minimized into the stage manager thumbnail.

**Comment by AI-Seppl** [comment]
I can reproduce the issue anytime, Rectangle 0.81 (87), Log attached.
[rectangleLog.txt](https://github.com/user-attachments/files/16426979/rectangleLog.txt)


I can specifically describe the scenario which consistently fails:

- You need to have more than one e.g. MS Edge windows open and visible, i.e. not tucked away in a Stage Manager stack
- Drag one of the two windows to top/bottom/middle left

This will ALWAYS cause the dragged window to get stacked into Stage Manager, which always resides on the left-hand side of the screen.
 
@rxhanson, can you reproduce it now?

**Comment by joshuajohncohen** [comment]
Hi,

I am also having this issue. For me, I am trying to split screen two Microsoft Word windows while having a third browser window open which is not being snapped. Latest Rectangle 0.85 (91), [rectangleLog.txt](https://github.com/user-attachments/files/18550946/rectangleLog.txt).

It only happens when I try to snap the windows to the left, where stage manager is, and it doesn't happen if I only have two MS Word windows open; I need two MS Word windows and one browser window.

**Comment by AI-Seppl** [comment]
IMHO, this is the case for anyone using Rectangle or Magnet, but I guess folks looking for sophisticated windows management options rarely use the mouse based snap functionality.

As I have neither seen a response on the issue nor a Rectangle version that addresses it, I tried other 3rd party windows management tools like Magnet and experienced the same.

At the very end, I am now happy with what‘s available natively since Sequoia. Though not that many options, it just works.
I guess I will never look back.

**Reply by rxhanson** [reply]
Sorry to hear that you've had a negative experience here, and apologies for not responding to your initial report. Unfortunately, this issue is due to the way that macOS works. As far as I know, the drop in the Stage Manager area in your scenario overrides any behavior that Rectangle or another third party would work. I think the only way to get the best of both worlds is to have a fix on Apple's side that would accommodate third parties, although it seems unlikely to me that this would be the resolution. If you're interested in coming back to Rectangle (which still has the benefit of being open source :), then at the very least another filing of this issue with Apple would still be helpful.

**Comment by pereljon** [comment]
Has anyone figured out how to resolve this issue? Dragging to the left incorporates the window into Stage Manager even though the Rectangle "preview" window outline is visible.

**Reply by rxhanson** [reply]
Thanks for asking. Since it's macOS behavior to absorb a dragged window into Stage Manager, a workaround here would likely be a hack of sorts to force the Stage Manager to hide when a window is being dragged. I have an idea on how to do that, but it's fairly involved and I would only consider implementing it if there were significant demand. If anyone wants to give it a go, let me know - I see no reason to turn down a good pull request with it.

**Reply by pereljon** [reply]
Ya. Figured that out so I'm using the left side of the bottom of the screen instead. Maybe the right function would be to disable the left side action if Stage Manager is enabled?

**Reply by rxhanson** [reply]
There is a slight snag with that, in that if Stage Manager deems that the dragged window should not be absorbed into the stage, then it does work fine to snap it. I agree, though, it's a bug to show it when it's not an option. I'll add this to my backlog.
